### PR TITLE
submission eventonica part 5 express js

### PR DESCRIPTION
**Pull request by Q-up**
Note to volunteers:
I am in the middle of transitioning my .html static file to the index.handlebars. All of index.js express code works on the server level in postman.

### CODE REVIEW

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+function validateInputUsers(user) {
+    const schema = {
+        name: Joi.string().min(3).required(), 
+        id: Joi.number().integer().min(5)
+    };
+    return Joi.validate(user, schema);
+};
+
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: Your REST APIs look great and your code is clear and concise!  I'd like to suggest naming the routes shorter and perhaps slightly more consistent.  For example, if the [domain] name of your server is eventonica.com, users would call the first two APIs by http://eventonica.com/api/er (to get all events) and http://eventonica.com/api/er/users (to get all users).  Note that this Eventonica server either returns information or enables the user to submit information.  Therefore, we can shorten the GET routes to:
/events and /users, i.e. http://eventonica.com/events ; http://eventonica.com/users

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+//display all events/users
+
+app.get('/api/er', (req, res) => {
+    res.send(events);
+});
+
+app.get('/api/er/users', (req, res) => {
+    res.send(users);
+});
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: Your REST APIs look great and your code is clear and concise! I'd like to suggest naming the routes shorter and perhaps slightly more consistent. For example, if the [domain] name of your server is eventonica.com, users would call the first two APIs by http://eventonica.com/api/er (to get all events) and http://eventonica.com/api/er/users (to get all users). Note that this Eventonica server either returns information or enables the user to submit information. Therefore, we can shorten the GET routes to:
/events and /users, i.e. http://eventonica.com/events ; http://eventonica.com/users

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    if (!event) return res.status(404).send('The event with the given ID was not found.');
+    res.send(event);
+});
+
+app.get('/api/er/users/:id', (req, res) => {
+    const user = users.find(u => u.id === parseInt(req.params.id)); 
+    if(!user) return res.status(404).send('The user with the given ID was not found.');;
+    res.send(user);
+});
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: Following the previous suggestion of route naming, GET by id would be:
/events/:id
/users/:id

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+// get event by keyword
+
+app.get('/api/er/search/:keyword', (req, res) => {
+    res.send(req.params.keyword);
+})
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: For 'get event by keyword', I have a couple of suggestions:
1. by path, /search/:keyword (this is as you have it currently)
2. by using query parameter, e.g. route would be GET /events
    and you would send back the event(s) that match the value of req.query.keyword
    and the caller of the API would need to do e.g.:  /events?keyword='volleyball'
    and in that case the value of req.query.keyword is 'volleyball'

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    const user = {
+        name: req.body.name,
+        id: Math.floor(Math.random()*90000) + 10000,
+        events: req.body.events
+    };
+    users.push(user);
+    res.send(user);
+});
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: Following the suggested route naming convention above, creating (aka POST) new events and users would be:
/events and /users

So now we would have routes that look like this in our API documentation or README:
GET /events
GET /users
GET /events/:id
GET /users/:id
GET /search/:keyword OR the same GET /events (with handling of req.query.keyword)
POST /events
POST /users

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+//     if (!event) res.status(404).send('The event with the given ID was not found.');
+// });
+
+//update an existing event
+app.put('/api/er/:id', (req, res) => {
+
+    //find event by id
+    const event = events.find(e => e.id === parseInt(req.params.id));
+    if (!event) return res.status(404).send('The event with the given ID was not found.');
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: excellent!

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+app.put('/api/er/:id', (req, res) => {
+
+    //find event by id
+    const event = events.find(e => e.id === parseInt(req.params.id));
+    if (!event) return res.status(404).send('The event with the given ID was not found.');
+    
+    //validate input
+    const {error} = validateInputEvents(req.body);
+    if(error) return res.status(400).send(error.details[0].message);
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: great to validate input!

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    if(error) return res.status(400).send(error.details[0].message);
+   
+    //update user
+    user.name = req.body.name,
+    // user.id = req.body.id,
+    user.events = req.body.events
+    //return the updated user 
+    res.send(user);
+});
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: Following the suggested route naming convention above, updating an existing event or user would be:
/events/:id, /users/:id

File: Eventonica-Part-5/eventonica-part-5/index.js [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+app.delete('/api/er/users/:id', (req, res) => {
+    const user = users.find(e => e.id === parseInt(req.params.id));
+    if (!user) return res.status(404).send('The user with the given ID was not found.');
+
+    const index = users.indexOf(user);
+    users.splice(index, 1);
+
+    res.send(user);
+});
+
```
Comment (iknockout) [inline Eventonica-Part-5/eventonica-part-5/index.js]: Following the suggested route naming convention above, deleting an existing event or user would be:
/events/:id, /users/:id

So finally we would have routes that look like this in our API documentation or README:
GET /events
GET /users
GET /events/:id
GET /users/:id
GET /search/:keyword OR the same GET /events (with handling of req.query.keyword)
POST /events
POST /users
PUT /events/:id
PUT /users/:id
DELETE /events/:id
DELETE /users/:id

**Comment by vegetabill** [comment]
@Q-up I marked this as 👍 from the review above. Feel feel to incorporate the feedback in later Parts.

**Pull request merged**